### PR TITLE
Add P6 update proposal draft.

### PR DESCRIPTION
**Pull request by abizjak**
## Purpose

Add protocol 6 update proposal.

## Checklist

- [ ] My code follows the style of this project.
- [ ] The code compiles without warnings.
- [ ] I have performed a self-review of the changes.
- [ ] I have documented my code, in particular the intent of the
      hard-to-understand areas.
- [ ] (If necessary) I have updated the CHANGELOG.

### CODE REVIEW

File: updates/P6.txt [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+sponsored transactions.
+The major update in protocol 6 however, is the new consensus protocol,
+Concordium BFT.
+
+Up to and including protocol 5, Concordium used a two-layer consensus
+design, with a proof of stake Nakamoto style consensus at the base
+layer with finalization on top which marked certain blocks as finalized.
+The idea being that the base Nakamoto layer works as long as more than
+1/2 of stake adheres to the protocol over long periods of time, but
+the finalization layer ensures speedy, explicit, finalization, in good
```
Comment (chrmatt) [inline updates/P6.txt:34]:
```suggestion
the finalization layer ensures speedy, explicit finalization, in good
```

File: updates/P6.txt [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+
+While this two-layer design did serve Concordium well during the initial
+phase, and is in theory robust, the two layer design is rather complex,
+and makes it challenging to optimize the consensus for both throughput and
+confirmation latency while maintaining security.
+
+In protocol 6 a new consensus protocol Concordium Byzantine Fault Tolerance
+(BFT) will be used instead. The protocol offers high transaction throughput
+and lower confirmation time because a block can be produced as soon as the
+previous block has been signed. The overall architecture remains the same.
```
Comment (chrmatt) [inline updates/P6.txt:45]: "certified" would probably be more precise than "signed" since one signature is not enough.

File: updates/P6.txt [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+There are still bakers and a subset of them, the finalizers. In contrast to
+how time was split into slots in existing consensus, Concordium BFT no
```
Comment (chrmatt) [inline updates/P6.txt:47]:
```suggestion
how time was split into slots in the existing consensus, Concordium BFT no
```

File: updates/P6.txt [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+The following behaviours are changed in protocol version 6.
+
+1. Concordium BFT takes effect. In particular this changes the format of
+   blocks, and the messages exchanged by the nodes.
+2. The election difficulty chain parameter is removed.
+3. There are new chain parameters
+   - the minimum timeout, and factors for increasing and decreasing round
+     timeout duration on successive round timeouts and finalizations,
+     respectively
+   - minimum block time, controlling minimum time between blocks in two
```
Comment (chrmatt) [inline updates/P6.txt:81]:
```suggestion
   - minimum block time, controlling the minimum time between blocks in two
```

File: updates/P6.txt [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+     parameter, revise the condition on being a finalizer. The relative
+     stake threshold is not interpreted relative to the total effective
+     stake of bakers. The minimum number of finalizers must be reached
+     before the relative stake threshold applies. Analogously, the number
+     of finalizers are capped.
+4. Block energy limit becomes an updatable parameter.
+5. Smart contract validation has the following changes
+   - Version 1 smart contracts can now use sign extension Wasm
+     instructions (i32.extend8_s, i32.extend16_s, i64.extend8_s,
+     i64.extend16_s, i64.extend32_s).
```
Comment (chrmatt) [inline updates/P6.txt:94]: It is not clear to me what that means. What do these extensions do?

File: updates/P6.txt [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+   - In version 1 smart contracts, the contract schema, and other custom
+     sections, no longer count towards the cost when executing contracts
+     for modules that were deployed in protocol 6.
+   - The behaviour of rollbacks in certain cases of reentrant contract
+     calls is corrected to conform to the intended semantics. In protocols
+     4 and 5 state changes might not be rolled back in case of failure of
+     nested, reentrant, contract calls.
```
Comment (chrmatt) [inline updates/P6.txt:101]:
```suggestion
     nested, reentrant contract calls.
```

File: updates/P6.txt [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+
+                               Abstract
+
+This document describes the changes in Concordium protocol version 6
+compared to protocol version 5.
+
+The protocol change has the following main effects
+- it changes the consensus protocol from an existing two-layer design
+  to a new Concordium BFT consensus
+- fixes the behaviour of state rollbacks in specific scenarios with
```
Comment (chrmatt) [inline updates/P6.txt:14]: I think Concordium uses American English, i.e., behavior. This also applies in other places below.

File: updates/P6.txt [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+
+1. The consensus change has no direct effects on the state of accounts and
+   contracts. The overall immediate effect of the new consensus will be, in
+   normal network conditions, that transaction confirmation times will have
+   much less variance.
+
+2. Behaviour of existing V1 contract instances may be affected in the
+   following scenarios
+   - the contract attempted to perform an operation with tags 5 or 6 the
+     execution would have triggered a runtime error, whereas in protocol 6
```
Comment (chrmatt) [inline updates/P6.txt:116]: Grammar? Should it be "if the contract..."? Generally quite long sentence that is hard to follow. It's also unclear to me what these tags are.

**Review by chrmatt — APPROVED** [review]
Looks good overall.

**Review by chportma — APPROVED** [review]

**Review by tschudid — APPROVED** [review]
Looks good to me. Christian's comments make sense.

**Review by concordium-cl — APPROVED** [review]

File: source/updates/P6.rst [hunk 9]
```diff
@@ -15,6 +15,9 @@ Protocol Update 6 - Concordium BFT consensus protocol, changes in Wasm validatio
 Specification
 =============
 
+.. include:: ../../updates/P6.txt
```
[the remark this answers is not in the document]
Comment (abizjak) [reply source/updates/P6.rst:18]: [YOLO.](https://www.youtube.com/watch?v=mLyOj_QD4a4)

File: updates/P6.txt [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+Builds on Protocol Version: 5
+
+                               Abstract
+
+This document describes the changes in Concordium protocol version 6
+compared to protocol version 5.
+
+The protocol change has the following main effects
+- it changes the consensus protocol from an existing two-layer design
+  to a new Concordium BFT consensus
```
[the remark this answers is not in the document]
Comment (abizjak) [reply updates/P6.txt:13]: Ok

File: updates/P6.txt (hunk 5 above)
Comment (abizjak) [reply updates/P6.txt:94]: I added a link to the Wasm spec/instructions.

The actual sign extension proposal only has a rather poor github repository to reference really.

File: updates/P6.txt (hunk 7 above)
Comment (abizjak) [reply updates/P6.txt:14]: True.

File: updates/P6.txt (hunk 8 above)
Comment (abizjak) [reply updates/P6.txt:116]: I revised a bit,

Tags are just operation ids. I don't want go into their details here, this is the level of details we used for previous documents.

File: updates/P6.txt [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+     respectively
+   - minimum block time, controlling minimum time between blocks in two
+     successive rounds
+   - minimum number of finalizers and maximum number of finalizers. These,
+     together with the existing finalizer relative stake threshold
+     parameter, revise the condition on being a finalizer. The relative
+     stake threshold is not interpreted relative to the total effective
+     stake of bakers. The minimum number of finalizers must be reached
+     before the relative stake threshold applies. Analogously, the number
+     of finalizers are capped.
```
[the remark this answers is not in the document]
Comment (abizjak) [reply updates/P6.txt:89]: Added
